farmfs/volume.py

```diff
@@ -291,20 +288,19 @@ def tree_patch(local_vol, remote_vol, delta):
     dst_blob = None
     src_blob = None
   if delta._mode == delta.REMOVED:
-    print "Apply", "Removing %s" % delta._path #TODO printing
-    ensure_absent(path)
-    # print "Apply", "Removing %s complete" % delta._path
+    return [(partial(ensure_absent, path), "Apply Removing %s" % delta._path)]
   elif delta._mode == delta.DIR:
-    print "Apply", "mkdir %s" % delta._path #TODO printing
-    ensure_dir(path)
+    return [(partial(ensure_dir, path), "Apply mkdir %s" % delta._path)]
   elif delta._mode == delta.LINK:
-    print "Apply", "mklink %s -> %s" % (delta._path, delta._csum) #TODO printing
+    #TODO i've complected tree diff and blob-store diff.
+    #TODO If order gets mixed up we might not copy data, or copy multiple times.
```

Comment:
This is a big problem. We need to change the diff metaphore to a blob diff as well as a tree diff. I don't think we can do diff apply piecemeal.